### ISSUE TITLE
cabal-2.4.0.0 Could not deserialize <repo>/root.json

### ISSUE THREAD

**Issue by Zq19491001** [haskell/cabal#8087]
**Describe the bug**
Even I use $ cabal new-update command , it also shows that Could not deserialize <repo>/root.json : Malformed.
**To Reproduce**
```
$ cabal new-update   or cabal v1-install  <any packages> ...
```

Please use version-prefixed commands (e.g. `v2-build` or `v1-build`) to avoid ambiguity.

**System information**
 - I use ubuntu 20.04
 - `cabal-2.4.0.0`, `ghc-8.6.5` versions

**Comment by fgaz** [comment]
Could you paste the exact output? Also could you try upgrading cabal to 3.6? 2.4 is no longer supported

**Comment by Zq19491001** [comment]
![image](https://user-images.githubusercontent.com/99885690/162444168-30cb6552-9064-4115-b88a-d033870727ce.png)

But I use ghcup to update cabal to 3.6.2.0 and ghc's version is 8.10.7 
cabal still fail.
![image](https://user-images.githubusercontent.com/99885690/162446369-83601618-86b7-4e18-8494-42f0f0ecc062.png)




By the way, I use cabal to install ogma, if you have free time ,the link is here:  https://github.com/nasa/ogma#table-of-contents
 
![image](https://user-images.githubusercontent.com/99885690/162446728-d820cea2-453c-46bd-8b0a-9cea2378f794.png)

**Comment by Mikolaj** [comment]
So cool! Though I wish the installation method was updated to new style cabal usage.

Anyway, please try removing ~/.cabal and ~/.ghc.

**Comment by Zq19491001** [comment]
Thanks ! So I remove  ~/cabal   file  (Using ghcup command: ghcup install ghc 8.10.7 to install ghc, there is no ~/ghc file), but the result is 
![image](https://user-images.githubusercontent.com/99885690/162454463-5dfddc4d-823e-4f41-b83f-32f1fe5b8e7d.png)

And cabal defaultly to build ~/.cabal  file again.

**Comment by Mikolaj** [comment]
Some blind shots: what is the filesystem type of this virtual machine partitions? Is it possible that somehow Windows line endings creep in? Are all relevant partitions mounted read-write? Are enough ports open to use curl (I guess yes, because ghcup works fine)?

@gbaz: does it look like a problem to connect to Hackage or corrupted transmission or corrupted deserialization of data from Hackage?

**Comment by Zq19491001** [comment]
Yes, I had seen #8031 issue before, so it is difficult for me to solve it.
the output of    find / -name "*root.json*"   is：
![image](https://user-images.githubusercontent.com/99885690/162461229-8c1ac011-20bf-45b4-a1dc-d419f89d8ffe.png)

but you know , when I use "apt-get install ghc cabal-install" in Ubuntu 18.04... , it will download cabal-install_1.24.0.0(I remember), that cabal-install version can work successfully. I do not know why

**Comment by Zq19491001** [comment]
Sir, I use following commands:
![image](https://user-images.githubusercontent.com/99885690/162463297-bba006ac-c715-4088-9aeb-40187fdd94f5.png)
because an error about  .yaml problems, the link is here: https://gitlab.haskell.org/haskell/ghcup-hs/-/issues/336

**Comment by Mikolaj** [comment]
Thank you for the listing and for the info ghcup and about cabal-install_1.24.0.0. A lot has change since then, so possibly we access the network differently than in 1.24.0.0.

So it seems there is no "root.json" file anywhere there. So it must be some internal cabal thing, probably when connecting to Hackage. Could you confirm with `cabal --version` and let us know if it turns out the invoked cabal is *not* 3.6 --- just in case.

Given that you need the workaround for ghcup, it's possible your machine has some ports blocked that are needed for connection to Hackage or your vm or VPN or Proxy blocks something. @gbaz: do we have any alternative ways to access Hackage?

**Comment by Zq19491001** [comment]
![image](https://user-images.githubusercontent.com/99885690/162470520-da1d3d59-9bf6-41c3-a587-18d0ec27b73a.png)
Thank you, Sir.

**Comment by Zq19491001** [comment]
Why my `~/.cabal/packages/hackage.haskell.org is  empty and there is no file named 'root.json'.
![image](https://user-images.githubusercontent.com/99885690/162563002-55ed54b6-7a7d-4224-b131-b0a8f59e93b8.png)




This is maybe wrong

**Comment by Mikolaj** [comment]
Indeed, I have that file:

/home/mikolaj/.cabal/packages/hackage.haskell.org/root.json

I think it's likely curl (if I remember right that's the method `cabal update` uses) does not work for you. I don't know if there are alternatives. Can you test if curl works for you? I bet google shows how to do it and perhaps has some workarounds for specific problems...

**Comment by Zq19491001** [comment]
I can use  curl https://hackage.haskell.org/root.json  and return 
![image](https://user-images.githubusercontent.com/99885690/162564040-47d1f13d-2c74-40b0-8d7a-62e4fd00274a.png)

Here is my curl version:
![image](https://user-images.githubusercontent.com/99885690/162564063-af02861e-a70c-419d-b0a1-6312b2738e52.png)

So could you show me that your curl version so that I can download it usefully.

**Comment by Mikolaj** [comment]
> I can use curl https://hackage.haskell.org/root.json and return

Oh, so it works for you? Then why it didn't seem to work when called inside ghcup? I'm confused. Perhaps you network permits curl to use the https protocol, but not the one used in ghcup and cabal (no idea which it is; http?)?

BTW, look what I've found:

    doc/cabal-project.rst:.. cfg-field:: http-transport: curl, wget, powershell, or plain-http

and here's the rendered version:

https://cabal.readthedocs.io/en/latest/cabal-project.html?highlight=http-transport

> Here is my curl version, it's very old:

```
~$ curl --version
curl 7.47.0 (x86_64-pc-linux-gnu) libcurl/7.47.0 GnuTLS/3.4.10 zlib/1.2.8 libidn/1.32 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets 
```

**Comment by Zq19491001** [comment]
Ok  thank you ,Sir. Old might be better, haha.

**Comment by Zq19491001** [comment]
Can you believe that I just uninstall my original curl and use " $sudo apt-get curl " to reinstall curl in my vm:
![image](https://user-images.githubusercontent.com/99885690/162573377-cade8f5b-e64d-4fc2-9ef4-cc9fc2ddc0b1.png)

Then all problems are missing. Thank you very much! So I think ghcup and cabal need to install corresponding curl version to prevent errors.

**Comment by jneira** [comment]
> So I think ghcup and cabal need to install corresponding curl version to prevent errors.

`cabal`  lets you change the `http-transport` if the default one is not working. In the actual cabal design required programs like `curl`, `ghc`, `hs2hc`, etc etc are not automatic downloaded and users are required to install, configure and update them.
That feature would add too much complexity as `cabal` was not designed to allow it in the first place.

So i dont think it something which would be implemented in a short or medium term. Maybe it would need better docs to make clear users the relation between cabal and other executables.

**Comment by Mikolaj** [comment]
Amazing! @robert-pearson, could you try any of the solutions we proposed here for #8031?

As @jneira explains, we are not in a position to install packages for people and bundling curl with ghcup or cabal would be cumbersome, but it would help to understand better what happened there. E.g., is that a know bug in curl 7.82? If so, perhaps we could mention in the docs that "if curl doesn't work (e.g., as in the case of the known buggy curl 7.82) and installing a curl version that works correctly is not an option (e.g., no administrator priviledges), then the `http-transport` option enables alternative methods of downloading Hackage data".
